src/pip/_internal/operations/install/wheel.py

```diff
@@ -600,7 +601,14 @@ def _generate_file(path, **kwargs):
             generated=generated,
             lib_dir=lib_dir)
     with _generate_file(record_path, **csv_io_kwargs('w')) as record_file:
-        writer = csv.writer(record_file)
+
+        # The mypy reveal_type record_file is different for
+        # Python 3 (typing.IO[Any]) and Python 2 (typing.BinaryIO),
+        # leading us to explicitly cast to typing.IO[Any] as a workaround
+        # for bad Python 2 behaviour
+        record_file_obj = cast('typing.IO[Any]', record_file)
```

Comment:
I believe you want `typing.IO[str]` (str being native string in both python2 and python3)

better maybe would be to change `csv_io_kwargs` to return a `Literal` such that mypy could infer the overload

Reply:
> I believe you want `typing.IO[str]` (str being native string in both python2 and python3)

I tried it and this works too.

> better maybe would be to change `csv_io_kwargs` to return a `Literal` such that mypy could infer the overload

Could you please expand on this. I found [typing.Literal](https://docs.python.org/3/library/typing.html#typing.Literal) in the docs and then modified the `csv_io_kwargs` to below and got rid of the `cast`, but the mypy assertions started failing again. Did you mean something like below, or did I miss something?

```python
from typing import Dict,  Literal
MODE = Literal['r', 'rb', 'w', 'wb']
def csv_io_kwargs(mode):
    # type: (MODE) -> Dict[str, Any]
    """Return keyword arguments to properly open a CSV file
    in the given mode.
    """
    if sys.version_info.major < 3:
        return {'mode': '{}b'.format(mode)}
    else:
        return {'mode': mode, 'newline': ''}
```

Reply:
the overloads are actually pretty complicated to get this correct

essentially what it boils down to is returning the right `Literal` -- here's a simpler example to illustrate

```python
if sys.version_info < (3,):
    @overload
    def mode(mode_s: Literal['r']) -> Literal['rb']: ...
    @overload
    def mode(mode_s: Literal['w']) -> Literal['wb']: ...
    def mode(mode_s: Union[Literal['r'], Literal['w']]) -> Union[Literal['rb'], Literal['wb']]:  # I think Literal['rb', 'wb'] is a shorthand for the union but I'm not at a treminal
        return '{}b'.format(mode_s)
else:
    ...  # python 3 left as an exercise
```

it's a little more difficult when you return a mapping though, you would probably need `TypedDict`

perhaps a more explicit design would be:

```python
if sys.version_info < (3,):
    def csv_read(*args: Any, **kwargs: Any) -> IO[str]:
        ...
    def csv_write(...) -> IO[str]:
        ...
else:
    ...
```

essentially the string format conditional is too clever for mypy and any of the portions which split the types by sys.version_info (which is what's actually happening) would teach it what's going on

Reply:
Thanks for the detailed explanation. Although in my personal opinion, I think that this level of complexity to define a type annotation might not be worth it, and it's easier for us to use explicitly `cast('IO[str],...)` as suggested before. 

I think we are going to drop Python 2 support after a few releases anyways, after which both the cast and the `if sys.version_info < (3,):` conditional will be removed.

Reply:
yeah I mean it would be more type-sound to do it that way instead of having `Dict[str, Any]`